Mathlib/Algebra/Category/ModuleCat/Differentials/Presheaf.lean

```diff
@@ -0,0 +1,188 @@
+/-
+Copyright (c) 2024 Joël Riou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou
+-/
+import Mathlib.Algebra.Category.ModuleCat.Presheaf
+import Mathlib.Algebra.Category.ModuleCat.Differentials.Basic
+
+/-!
+# The presheaf of differentials of a presheaf of modules
+
+In this file, we define the type `M.Derivation φ` of derivations
+with values in a presheaf of `R`-modules `M` relative to
+a morphism of `φ : S ⟶ F.op ⋙ R` of presheaves of commutative rings
+over categories `C` and `D` that are related by a functor `F : C ⥤ D`.
+We formalize the notion of universal derivation.
+
+Geometrically, if `f : X ⟶ S` is a morphisms of schemes (or more generally
+a morphism of commutative ringed spaces), we would like to apply
+these definitions in the case where `F` is the pullback functors from
+open subsets of `S` to open subsets of `X` and `φ` is the
+morphism $O_S ⟶ f_* O_X$.
+
+In order to prove that there exists a universal derivation, the target
+of which shall be called the presheaf of relative differentials of `φ`,
+we first study the case where `F` is the identity functor.
+In this case where we have a morphism of presheaves of commutative
+rings `φ' : S' ⟶ R`, we construct a derivation
+`DifferentialsConstruction.derivation'` which is universal.
+Then, the general case (TODO) shall be obtained by observing that
+derivations for `S ⟶ F.op ⋙ R` identify to derivations
+for `S' ⟶ R` where `S'` is the pullback by `F` of the presheaf of
+commutative rings `S` (the data is the same: it suffices
+to show that the two vanishing conditions `d_app` are equivalent).
+
+-/
+
+universe v u v₁ v₂ u₁ u₂
+
+open CategoryTheory
+
+variable {C : Type u₁} [Category.{v₁} C] {D : Type u₂} [Category.{v₂} D]
+
+namespace PresheafOfModules
+
+variable {S : Cᵒᵖ ⥤ CommRingCat.{u}} {F : C ⥤ D} {S' R : Dᵒᵖ ⥤ CommRingCat.{u}}
+   (M N : PresheafOfModules.{v} (R ⋙ forget₂ _ _))
+   (φ : S ⟶ F.op ⋙ R) (φ' : S' ⟶ R)
```

Comment:
I'm still not fully convinced that our main API should built around a specific type of sheaf homs. Can't we just pullback `M` along `F` when necessary?

Reply:
This is not a "specific type" of morphisms. This is *more general* than "any" `S ⟶ R`.

Reply:
But if you are developing material (e.g. `relativeDifferentials'`) for abstract `S ⟶ R` below. I don't see how you can apply these back to morphisms of the form `S ⟶ F.op ⋙ R` (or $F^{-1}\mathcal{O}_S \to \mathcal{O}_R$) that allows use of this general API. In this sense, the material you are developing below is more restricted. 

I really believe that it would be cleaner if we can deal with pushforward-pullbacks separately later.

Another concern is, when you are viewing an abstract `(φ : S ⟶ R)` as a `(φ : S ⟶ (1 _).op ⋙ R)`, it is abusing the underlying defeq of `(1 _)` and `op` etc, and I'm not sure how happy lean is with this when the `R` and `S` and the category are more complex.

Reply:
Part of my point is that when we want to apply this to a morphism of schemes, then such a morphism is described as a morphism of topological spaces (which gives a certain functor `F` on categories of opens subsets) and a map `φ : Y.presheaf ⟶ base _* X.presheaf` which fits precisely in the API I am suggesting. We could define $Omega^1$ by using the adjoint map `base ^* Y.(pre)sheaf ⟶  X.(pre)sheaf`, but developing the API would be ackward because the pullback as a presheaf is not the same thing as the pullback as a sheaf.

The API you suggest is the *particular* case `Derivation'` in my PR. Using the more general API I suggest is a way to anticipate some issues when we want to sheafify.

Reply:
We wouldn't need to sheafify the sheaf of rings though, so "pullback as a presheaf is not the same thing as the pullback as a sheaf" is a non-issue as we will work with the pullback sheaf all along.

Maybe I should first ask how you plan to get `relativeDifferentials` from `relativeDifferentials'`.

Reply:
> Maybe I should first ask how you plan to get `relativeDifferentials` from `relativeDifferentials'`.

This is explained in the docstring of the file.

Reply:
Aah okay fair enough.

Reply:
Also, the equivalence of derivations for `S ⟶ F.op ⋙ R` and for `F^* S ⟶ R` is of independent interest, and it can be stated only if the more general definitions are available (or if we had used absolute derivations).

Reply:
Thanks for your understanding :-)